### PR TITLE
K8s infrastructure provisioner

### PR DESCRIPTION
**Pull request by jpdhas**
Can the creation of the cluster be reviewed on its own please?

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,31 @@
+FROM alpine:3.6
```
Comment (sbuliarca) [inline Dockerfile]: We have already an image that has the k8s cli utilities: https://github.com/Financial-Times/k8s-cli-utils.
It is used by Jenkins to communicate with the clusters.
It is a little out of date, but maybe we can use it.

File: sh/generate_cluster_config.sh [hunk 2]
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env bash
+
+TEMPLATE_FILE=cluster-template.yaml
```
Comment (sbuliarca) [inline sh/generate_cluster_config.sh:3]: I was thinking of using Ansible for its templating engine, and ditch this sh. Using ansible has a few advantages, as it gives us control loops that we can use, for example for iterating over a list of additional tags.

File: sh/provision.sh [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+# Generate the cluster config file from the template
+if [ ! -f ${CONFIGS_FOLDER}/${CLUSTER}.properties ]; then
+    echo "Cluster config file not found"
+else
+    /bin/sh generate_cluster_config.sh ${CLUSTER}
+fi
+
+# Create the TLS assets for the cluster
+kube-aws render credentials --generate-ca
```
Comment (sbuliarca) [inline sh/provision.sh:17]: Will we use a new CA key for each cluster? We were thinking of sharing at least between publishing and delivery in the same region.

File: sh/provision.sh [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+export AWS_ACCESS_KEY_ID="${AWS_ACCESS_KEY_ID}"
+export AWS_SECRET_ACCESS_KEY="${AWS_SECRET_ACCESS_KEY}"
+
+# Generate the cluster config file from the template
+if [ ! -f ${CONFIGS_FOLDER}/${CLUSTER}.properties ]; then
+    echo "Cluster config file not found"
+else
+    /bin/sh generate_cluster_config.sh ${CLUSTER}
+fi
+
```
Comment (sbuliarca) [inline sh/provision.sh:15]: We should also cleanup the credentials folder before generating them.

File: sh/provision.sh [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+esac
+
+# Validate the config
+kube-aws validate --s3-uri s3://${PROVISIONER_BUCKET}
+
+# Create the stack
+kube-aws up --s3-uri s3://${PROVISIONER_BUCKET}
+
+# Tag the subnets with the cluster value
+for subnet in $(aws ec2 describe-subnets --filters Name=vpc-id,Values=${VPC} --query Subnets[*].SubnetId --region=${REGION} --output text); do
```
Comment (sbuliarca) [inline sh/provision.sh:57]: Nice !

**Comment by sbuliarca** [comment]
This is a good start.
For UPP we'll need additional stuff like EBSs volumes for mongos & kafka, S3 bucket for image publish, etc.

File: sh/generate_cluster_config.sh (hunk 2 above)
Comment (jpdhas) [reply sh/generate_cluster_config.sh:3]: Ansible is good when it comes to provisioning infrastructure. And my personal opinion is that is should not be meant to be used for things that it is not meant to do 🙈 
But I am open to discussion about this. And I did try to use ansible. But couldn't figure out a way to edit files. Would be interested to learn how to do it 😄

**Pull request closed without merging**